Feature/incrwmean

MatthewCochrane opened the pull request:
Resolves #45  .

<!--lint disable first-heading-level-->

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [x] Read and understand the [Code of Conduct][code-of-conduct].
-   [x] Read and understood the [licensing terms][license].
-   [x] Searched for existing issues and pull requests **before** submitting this pull request.
-   [x] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [x] Rebased onto latest `develop`.
-   [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

-   Implements an incremental weighted mean

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #45 

## Questions

> Any questions for reviewers of this pull request?

Do you agree that the weight should default to 1.0?  then `incrwmean` with one parameter is equivalient to `incrmean` (though slower).

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

I'm not sure that other incremental stat functions have typescript definitions.  Didn't see why I shouldn't add one for this though.  Let me know if there's some reason I shouldn't.

I wasn't aware of a more efficient way to implement this algorithm, I don't believe there's a shortcut like for `incrmean`.

* * *

@stdlib-js/reviewers

<!-- <links> -->

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

<!-- </links> -->

MatthewCochrane committed:
Add readme

MatthewCochrane committed:
Add examples

MatthewCochrane committed:
Add package.json

MatthewCochrane committed:
Add implementation

MatthewCochrane committed:
Add benchmarks

MatthewCochrane committed:
Add tests and add weight default of 1.0

MatthewCochrane committed:
Update examples to reflect default weight parameter

MatthewCochrane committed:
Add repl

MatthewCochrane committed:
Add typescript definition

kgryte commented:
@MatthewCochrane This is awesome! For an incremental algorithm which only relies on maintaining the weighted mean and the sum of weights, see [here](http://people.ds.cam.ac.uk/fanf2/hermes/doc/antiforgery/stats.pdf).

Re: default weight. Personally, I would require a weight to always be provided. Meaning, we would not support a default weight. Otherwise, when reading code which uses `incrwmean`, one has to rely on "out-of-band" knowledge to know what the default weight is (a) and (b) if one is unfamiliar with the `incrwmean` API, it is not clear if the "missing" weight is intentional or not. Accordingly, my recommendation is that the API not be variadic.

Re: TypeScript files. Yes! We only recently started adding declaration files and have yet to add declaration files to the `incr/*` packages. WIP. :) 

Thanks for working on this!

MatthewCochrane committed:
Make weight parameter required

MatthewCochrane committed:
Modify incremental algorithm
Used algorithm from here http://people.ds.cam.ac.uk/fanf2/hermes/doc/antiforgery/stats.pdf

MatthewCochrane commented:
Not a problem, happy to work on it!

Yep, that's fair.  I'm not sure I'd call it 'out of band knowledge' since it's in the jsdoc header, and most (sane) people use editors with documentation lookup these days.  Though the saving would be quite small and you're right that it's probably worth the small amount of extra verbosity to save potential misinterpretations. I removed the default, as suggested :).

Thanks, for that link.  I've updated the algorithm to use the one from the paper.  It benchmarks very slightly faster than the one I had implemented.

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/README.md:
```diff
@@ -0,0 +1,129 @@
+
+The [weighted arithmetic mean][weighted-arithmetic-mean] is defined as
+
+<!-- <equation class="equation" label="eq:arithmetic_mean" align="center" raw="{\bar {x}}={\frac {\sum \limits _{i=1}^{n}w_{i}x_{i}}{\sum \limits _{i=1}^{n}w_{i}}}" alt="Equation for the arithmetic mean."> -->
```
```suggestion
<!-- <equation class="equation" label="eq:weighted_arithmetic_mean" align="center" raw="\bar {x}= \frac{\sum_{i=1}^{n} w_{i}x_{i}}{\sum_{i=1}^{n}w_{i}}" alt="Equation for the weighted arithmetic mean."> -->
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/README.md:
```diff
@@ -0,0 +1,129 @@
+
+<!-- <equation class="equation" label="eq:arithmetic_mean" align="center" raw="{\bar {x}}={\frac {\sum \limits _{i=1}^{n}w_{i}x_{i}}{\sum \limits _{i=1}^{n}w_{i}}}" alt="Equation for the arithmetic mean."> -->
+
+<div class="equation" align="center" data-raw-text="{\bar {x}}={\frac {\sum \limits _{i=1}^{n}w_{i}x_{i}}{\sum \limits _{i=1}^{n}w_{i}}}" data-equation="eq:arithmetic_mean">
```
```suggestion
<div class="equation" align="center" data-raw-text="\bar {x}= \frac{\sum_{i=1}^{n} w_{i}x_{i}}{\sum_{i=1}^{n}w_{i}}" data-equation="eq:weighted_arithmetic_mean">
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/README.md:
```diff
@@ -0,0 +1,129 @@
+
+#### incrwmean()
+
+Returns an accumulator `function` which incrementally computes an [weighted arithmetic mean][weighted-arithmetic-mean].
```
```suggestion
Returns an accumulator `function` which incrementally computes a [weighted arithmetic mean][weighted-arithmetic-mean].
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/repl.txt:
```diff
@@ -0,0 +1,37 @@
+    Returns an accumulator function which incrementally computes a weighted
+    arithmetic mean.
+
+    If provided parameter(s), the accumulator function returns an updated
```
```suggestion
    If provided arguments, the accumulator function returns an updated
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/repl.txt:
```diff
@@ -0,0 +1,37 @@
+    arithmetic mean.
+
+    If provided parameter(s), the accumulator function returns an updated
+    weighted mean. If not provided a value, the accumulator function returns
```
```suggestion
    weighted mean. If not provided arguments, the accumulator function returns
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/repl.txt:
```diff
@@ -0,0 +1,37 @@
+    If provided `NaN` or a value which, when used in computations, results in
+    `NaN`, the accumulated value is `NaN` for all future invocations.
+
+    The accumulator accepts two parameters, the first is the value and the
```
```suggestion
    The accumulator accepts two arguments: the first is the value and the
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/types/index.d.ts:
```diff
@@ -0,0 +1,52 @@
+*
+* @returns accumulator function
+*
+* @example* var incrwmean = require( '@stdlib/stats/incr/wmean' );
```
```suggestion
* @example
* var incrwmean = require( '@stdlib/stats/incr/wmean' );
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/types/index.d.ts:
```diff
@@ -0,0 +1,52 @@
+* mu = accumulator();
+* // returns 2.5
+*/
+declare function incrwmean(): ( x: number, w: number ) => number | null;
```
I think declaring a type definition for the return value is probably better here and then using that type definition for the return value. The reason being that, by creating the type definition, we are able to explicitly document the accumulator parameters and return values.

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/types/test.ts:
```diff
@@ -0,0 +1,39 @@
+	incrwmean(); // $ExpectType (x: number, w: number) => number | null
+}
+
+// The compiler throws an error if the function is any arguments
```
```suggestion
// The compiler throws an error if the function is provided arguments...
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/docs/types/test.ts:
```diff
@@ -0,0 +1,39 @@
+	incrwmean( [] ); // $ExpectError
+	incrwmean( {} ); // $ExpectError
+	incrwmean( ( x: number ): number => x ); // $ExpectError
+}
```
We should also include tests for the returned accumulator function.

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -0,0 +1,74 @@
+* var mu = accumulator();
+* // returns null
+*
+* var mu = accumulator( 2.0, 1.0 );
```
```suggestion
* mu = accumulator( 2.0, 1.0 );
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/index.js:
```diff
@@ -0,0 +1,54 @@
+* var mu = accumulator();
+* // returns null
+*
+* var mu = accumulator( 2.0, 1.0 );
```
```suggestion
* mu = accumulator( 2.0, 1.0 );
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -0,0 +1,74 @@
+	return accumulator;
+
+	/**
+	* If provided a value, the accumulator function returns an updated weighted mean. If not provided a value, the accumulator function returns the current weighted mean.
```
```suggestion
	* If provided arguments, the accumulator function returns an updated weighted mean. If not provided arguments, the accumulator function returns the current weighted mean.
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -0,0 +1,74 @@
+	* @private
+	* @param {number} [x] - new value
+	* @param {number} [w] - new value's weight
+	* @returns {(number|null)} weighted mean value or null
```
```suggestion
	* @returns {(number|null)} weighted mean or null
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/lib/main.js:
```diff
@@ -0,0 +1,74 @@
+			return mu;
+		}
+		wSum += w;
+		mu += ( (w / wSum) * (x - mu) );
```
```suggestion
		mu += (w / wSum) * (x - mu);
```

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/test/test.js:
```diff
@@ -0,0 +1,120 @@
+	t.end();
+});
+
+tape( 'if null is passed as a weight, the accumulator function returns NaN', function test( t ) {
```
This test can be removed.

kgryte commented on lib/node_modules/@stdlib/stats/incr/wmean/test/test.js:
```diff
@@ -0,0 +1,120 @@
+
+	acc = incrwmean();
+	acc();
+	t.ok( isNaN( acc(2.0) ), 'returns NaN' );
```
Use the `@stdlib/math/base/assert/is-nan` utility. `isNaN` is historically problematic in JavaScript.

kgryte reviewed (changes requested):
@MatthewCochrane Thanks for updating! Left some comments and suggestions. Once resolved, this should be ready for merge!

kgryte committed:
Apply suggestions from code review

Co-Authored-By: MatthewCochrane <matthew.cochrane.eng@gmail.com>

MatthewCochrane committed:
Update tests and typescript definition

MatthewCochrane committed:
Update typescript definition

MatthewCochrane committed:
Add typescript accumulator tests

kgryte reviewed (approved):
@MatthewCochrane Looks great! Thanks!
